client/my-sites/site-settings/redirect-to/index.jsx

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import page from 'page';
+import { flowRight } from 'lodash';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import { getSelectedSiteId, getSelectedSiteSlug } from 'state/ui/selectors';
+import { isJetpackSite } from 'state/sites/selectors';
+import { isSiteAutomatedTransfer } from 'state/selectors';
+
+const redirectNonJetpack = redirectRoute => WrappedComponent => {
+	class RedirectNonJetpack extends Component {
+		static propTypes = {
+			redirectRoute: PropTypes.string,
+			// Connected props
+			siteIsAtomic: PropTypes.bool,
+			siteIsJetpack: PropTypes.bool,
+			siteSlug: PropTypes.string
+		}
+
+		componentDidMount() {
+			this.redirectIfNoAccess();
+		}
+
+		componentDidUpdate() {
+			this.redirectIfNoAccess();
+		}
+
+		redirectIfNoAccess() {
+			if ( this.props.siteIsJetpack === false || this.props.siteIsAtomic ) {
+				this.redirectToDiffSite();
+			}
+		}
+
+		redirectToDiffSite = () => {
+			const { siteSlug } = this.props;
+
+			if ( siteSlug ) {
```

Comment:
This logic would be unnecessary for cases where we've specified a custom `redirectRoute`. You should not assert that any route that we redirect to will end with the site slug. Even if I want the custom `redirectRoute` to contain the site slug, I'd expect to have to build myself it prior to passing it to the HoC.

So, what I'd do here instead is:

* If we have a custom `redirectRoute`, just redirect to it, without caring about a site slug.
* If we use the default `redirectRoute`, append the `siteSlug` to it (if it's loaded), and then redirect to the result route.

This will benefit in several ways:
* Whoever uses the HoC will have more freedom for choosing what route to redirect to.
* If the site slug (or another site related data) is not contained in the custom redirect route, the HoC will be able to redirect instantly, instead of waiting for the sites to be loaded.

Please, let me know if that doesn't make any sense. 😉

Reply:
@tyxla, yes, makes sense! 
This and the following PRs made me wonder whether it is flexible enough to have both in-and output routes be the same.

Reply:
Thanks again @tyxla for your second to none attention to detail! 

All improvements are included now. 
Re this last part: IMHO it makes perfect sense to improve the logic here. I added one extra detail to it. `redirectRoute` is used as-is when it contains A site slug (any custom `siteSlug` created prior to passing the route to the HoC). In the absence of such, it will append the current `siteSlug` (if loaded) to the route. There is still freedom of choice here, but the transition is not broken by asking for a choice of site when no `siteSlug` was present. 

Please feel free to correct me if I'm missing an important point. Thank you for last last-minute look :) Hopefully now it is good to go 🤞

Reply:
I don't think we got that one right. Actually, I might have been a little unclear last time - sorry if that is the case.

So, with your current implementation it will not work well in some cases:

* If the user has specified a redirect route but the site slug is not included in that redirect route, we might unexpectedly alter their preferred redirect route. That should not happen - that's why I recommended that if the user has specified a custom redirect route, we should just redirect to that route, without modifying it in any way. 
* If the user has not specified a redirect route, and tries to access the page with a clean Redux store, the current site slug might not be loaded yet, so the user will be redirected to `/settings/general` without a site slug. It would be better to wait for the current site slug to be loaded and redirect to `/settings/general/:siteSlug` instead.

That said, instead of:

```
redirect = () => {
	const { siteSlug } = this.props;

	let route = '';
	if ( redirectRoute ) {
		const getSiteSlug = ( redirectRoute.split( '/' ) ).slice( -1 )[ 0 ];
		route = getSiteSlug ? redirectRoute : redirectRoute + siteSlug;
		page( route );
	} else {
		route = siteSlug ? '/settings/general/' + siteSlug : '/settings/general/';
		page( route );
	}
};
```

we might want to refactor it this way:

```
redirect = () => {
	const { redirectRoute, siteSlug } = this.props;

	if ( redirectRoute ) {
		return page( redirectRoute );
	}

	if ( siteSlug ) {
		page( '/settings/general/' + siteSlug );
	}
};
```

so it would support the cases I mentioned above.

Disclaimer: I haven't tested that code, just drafted it quickly here, but I hope you get the idea. Let me know if that makes sense 😉 

I think we should be good to go after we address that one. 👍

Reply:
@tyxla, this is all true. Thanks a lot!

I saw 1), yet did not see it as an undesirable behaviour. I actually thought we should fill up the Slug assuming it is a mistake/omission, thus the reg. expression in there. 
Why would we want to grant so much flexibility with a 'security' redirection? 🤔 

Re 2), yes, yes, yes 👍

Reply:
> Why would we want to grant so much flexibility with a 'security' redirection?

I understand why that's dubious. Well, you said it - flexibility - that's all what it is about, that's why we provide the `redirectRoute` prop to the component - because we want to provide some flexibility, and we should rely that the code using the component will pass a proper route. One can easily pass a wrong route even if the site slug is correct, right? We can't mitigate all of those mistakes. That's why I think it's best to rely on the wrapping component to pass the right route.